### PR TITLE
Add helm chart for operator based implementation

### LINKED ISSUE #268: Add helm chart for operator
With migration from controller to operator, we removed the helm chart out of the operator. Add helm chart back again since it's required for ease of use.

### PR DESCRIPTION
**Pull request by kahootali**
Closes #268

### CODE REVIEW

**Comment by stakater-user** [comment]
@kahootali Image is available for testing. `docker pull stakater/IngressMonitorController:SNAPSHOT-PR-315-ecd849e5`

**Comment by kahootali** [comment]
Somehow the [Action Run](https://github.com/stakater/IngressMonitorController/runs/2491441090?check_suite_focus=true) didn't run the [step](https://github.com/stakater/IngressMonitorController/pull/315/files#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cR36) I added in PR workflow

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -33,6 +33,11 @@ jobs:
         curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.26.0
         golangci-lint run --timeout=10m ./...
 
+    - name: Helm Lint
+      run: |
+        cd deploy/chart/ingressmonitorcontroller
```
Comment (ahmedwaleedmalik) [inline .github/workflows/pull_request.yml:38]: `helm lint deploy/chart/ingressmonitorcontroller`

**Comment by ahmedwaleedmalik** [comment]
@kahootali Thanks for the contribution. It didn't work because in case of a PR it uses upstream pipelines instead of the pipelines defined in the PR. For security reasons, for modre info: 
https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target

File: .github/workflows/push.yml [hunk 2]
```diff
@@ -135,6 +136,21 @@ jobs:
           VERSION: ${{ steps.generate_operator_tag.outputs.new_tag }}
         run: make update-resources-ci
 
+      # Publish helm chart
```
Comment (ahmedwaleedmalik) [inline .github/workflows/push.yml:139]: Please add step for helm lint here as well.

File: deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  - watch
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - daemonsets
+  - replicasets
+  - statefulsets
+  verbs:
+  - "*"
```
Comment (ahmedwaleedmalik) [inline deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml:44]: Are you sure we need to assign open access for these resources?

**Comment by stakater-user** [comment]
@kahootali Image is available for testing. `docker pull stakater/IngressMonitorController:SNAPSHOT-PR-315-cc2a7f57`

File: deploy/chart/ingressmonitorcontroller/templates/role.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  - watch
+- apiGroups:
+  - apps
+  resources:
+  - deployments
+  - daemonsets
+  - replicasets
+  - statefulsets
+  verbs:
+  - "*"
```
Comment (ahmedwaleedmalik) [inline deploy/chart/ingressmonitorcontroller/templates/role.yaml:44]: I think we need to restrict actions here.

File: deploy/chart/ingressmonitorcontroller/values.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+global:
+# labels added on all components of chart in addition to some default labels
+  labels: {}
+
+# if empty, IMC will look in all namespaces, 
+# you can restrict to single namespace
```
Comment (ahmedwaleedmalik) [inline deploy/chart/ingressmonitorcontroller/values.yaml:6]: We can specify multiple namespaces as well by adding comma separated list.

**Review by ahmedwaleedmalik — CHANGES_REQUESTED** [review]

File: deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml (hunk 3 above)
Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml:44]: just copied the role from `deploy/` folder, do you think if only `list,get,watch` will work?

File: deploy/chart/ingressmonitorcontroller/values.yaml (hunk 5 above)
Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/values.yaml:6]: ok, will try to fix this

File: deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml (hunk 3 above)
Comment (ahmedwaleedmalik) [reply deploy/chart/ingressmonitorcontroller/templates/clusterrole.yaml:44]: Yeah that should be it. I think `watch` won't be required now either.

File: deploy/chart/ingressmonitorcontroller/values.yaml (hunk 5 above)
Comment (ahmedwaleedmalik) [reply deploy/chart/ingressmonitorcontroller/values.yaml:6]: You just need to fix the comment, nothing else.

Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/values.yaml:6]: wouldn't it be better to create role & rolebinding for all the namespaces mentioned?

Comment (ahmedwaleedmalik) [reply deploy/chart/ingressmonitorcontroller/values.yaml:6]: Yes, if you can that would be awesome :) In case if a person specifies namespace and the flag is not empty then we need to create `role, rb` instead of `clusterRole, CRB`

Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/values.yaml:6]: yeah, thinking of doing that

File: .github/workflows/pull_request.yml (hunk 1 above)
Comment (kahootali) [reply .github/workflows/pull_request.yml]: updated

File: .github/workflows/push.yml (hunk 2 above)
Comment (kahootali) [reply .github/workflows/push.yml:143]: done

File: deploy/chart/ingressmonitorcontroller/values.yaml (hunk 5 above)
Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/values.yaml]: done, now if a user adds multiple namespaces comma separated, chart will create Role, RB in all those namespaces

**Comment by stakater-user** [comment]
@kahootali Image is available for testing. `docker pull stakater/IngressMonitorController:SNAPSHOT-PR-315-8d430df8`

File: .github/workflows/pull_request.yml [hunk 6]
```diff
@@ -20,7 +20,8 @@ jobs:
     - name: Check out code
       uses: actions/checkout@v2
       with:
-        ref: ${{github.event.pull_request.head.sha}}
+        fetch-depth: 0
+        ref: ${{ github.head_ref }}
```
Comment (ahmedwaleedmalik) [inline .github/workflows/pull_request.yml:24]: Please revert this change `ref: ${{ github.head_ref }}`

File: deploy/chart/ingressmonitorcontroller/templates/role.yaml [hunk 7]
```diff
@@ -0,0 +1,86 @@
+{{- if and ( .Values.rbac.create ) ( ne .Values.watchNamespaces "") }}
+{{- range .Values.watchNamespaces | split "," }}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ template "name" $ }}-role
+  namespace: {{ . | quote }}
```
Comment (ahmedwaleedmalik) [inline deploy/chart/ingressmonitorcontroller/templates/role.yaml:7]: Remove `quote` and add `trim`. So, if a user specifies `watchNamespaces` like "test, test-a,test" it would still work perfectly fine.

File: deploy/chart/ingressmonitorcontroller/templates/rolebinding.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+{{- if and ( .Values.rbac.create ) (ne .Values.watchNamespaces "") }}
+{{- range .Values.watchNamespaces | split "," }}
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: {{ template "name" $ }}-role-binding
+  namespace: {{ . | quote }}
```
Comment (ahmedwaleedmalik) [inline deploy/chart/ingressmonitorcontroller/templates/rolebinding.yaml:7]: Remove `quote` and add `trim`

**Review by ahmedwaleedmalik — CHANGES_REQUESTED** [review]
Just few minor changes requested

**Comment by stakater-user** [comment]
@kahootali Yikes! You better fix it before anyone else finds out! [Build](https://github.com/stakater/IngressMonitorController/commit/3c189eec8d196979140f77d5fdd5e151ac29131e/checks) has Failed!

File: deploy/chart/ingressmonitorcontroller/templates/role.yaml (hunk 7 above)
Comment (kahootali) [reply deploy/chart/ingressmonitorcontroller/templates/role.yaml]: done

**Comment by kahootali** [comment]
The pipeline failed due to a failed test, can anyone from the team check or re-run the pipeline?

**Comment by stakater-user** [comment]
@ahmedwaleedmalik Image is available for testing. `docker pull stakater/IngressMonitorController:SNAPSHOT-PR-315-3c189eec`

**Review by ahmedwaleedmalik — APPROVED** [review]
LGTM
